source/extensions/filters/network/http_connection_manager/dependency_manager.h

```diff
@@ -21,11 +21,16 @@ class DependencyManager {
   /**
    * Register each filter in an http filter chain, using name and dependencies
    * from the filter factory. Filters must be registered in decode path order.
+   *
+   * Currently, filters with no dependencies required or provided have no
+   * bearing on the validity of a filter chain, and are silently skipped.
    */
   void registerFilter(
       const std::string& filter_name,
       const envoy::extensions::filters::common::dependency::v3::FilterDependencies& dependencies) {
-    filter_chain_.push_back({filter_name, dependencies});
+    if (dependencies.ByteSizeLong() != 0) {
```

Comment:
I think passing an optional here would be clearer, this seems like an odd way to check whether there are any dependencies specified

Reply:
OK so the reason I would argue for keeping as-is: The API for this class is designed for someone to register filters in order and then check if its valid afterwards. I feel like skipping registration for empty configs is an implementation detail, and I don't want callers of registerFilter to have to think about whether a dependency is relevant or not. 

I can also imagine a case where we actually do want to register filters with empty FilterDependencies, e.g. maybe there's a future where we encode logic like "this filter must be second in the chain" as part of DependencyManager. And lastly when I add the encode path, we could potentially record maintain the decode/encode path separately.

Reply:
I just think the size check is a bit strange, since you could do weird stuff like set unknown fields on the proto and that would affect this check. Maybe just check the size of the encode/decode dependencies instead?

Reply:
Yeah I can see why the size check is strange, but this was actually the recommended guidance for checking if a proto is empty (i.e. default). The alternative is to check each for presence of each of the four sub-fields independently, which is a lot more brittle. I think if we were to add new fields to the dependencies proto it's a good thing that this check will be affected (and tests will break), and if new fields are added with a default value that won't affect the byteSizeLong(). I left a comment to clarify, what do you think?

Also oops I didn't commit my change before pushing, the other comment should be fixed now.

Reply:
I realize this is how you'd check for the default proto, I think I just didn't like treating the default proto special here :) 

I don't feel that strongly about it though (unless someone in @envoyproxy/maintainers disagrees), so this is probably fine

>I can also imagine a case where we actually do want to register filters with empty FilterDependencies, e.g. maybe there's a future where we encode logic like "this filter must be second in the chain" as part of DependencyManager.

If we're trying to avoid copying this around it seems like it won't be good to increase the scope here? Maybe we should be trying to avoid the cost here in another way? Can we just store the unique ptrs instead so we avoid copies?

Reply:
@htuch to weigh in

re:the scope, I was just talking about a hypothetical where we'd be expanding the scope, I think what we have right now is simple enough to manage without unique pointers, I don't really expect these to be complicated enough for the copy to be that costly.

Mostly I'm just concerned about clarity of the API – one other approach could be to split it into two functions, `registerFilter` and `registerFilterIfNotEmpty`/`maybeRegisterFilter`. Not sure if it's necessary but that could be enough to keep the API clean.

Reply:
Yeah I don't think the current usage requires any kind of optimization, I'm just working with the concerned raised by @alyssawilk and it seemed like we could avoid copies overall that would be better as the scope increases and more filters report some kind of dependency.

Do we have an understanding of what exactly the perf implications are here? If this just happens once or twice (upgrade chain) for HCM load this seems like it probably doesn't matter?

Reply:
If this only happens at HCM registration, I'm thinking it doesn't really matter a great deal. The only situation I worry deeply is the one that @alyssawilk points out when we have to do anything per request or per route. Listener scalability is challenged for other reasons and not something being generally pursued in the community.